docs/src/pages/demos/lists/SimpleList.tsx

```diff
@@ -0,0 +1,62 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { createStyles, Theme, withStyles, WithStyles } from '@material-ui/core/styles';
+import List from '@material-ui/core/List';
+import ListItem, { ListItemProps } from '@material-ui/core/ListItem';
+import ListItemIcon from '@material-ui/core/ListItemIcon';
+import ListItemText from '@material-ui/core/ListItemText';
+import Divider from '@material-ui/core/Divider';
+import InboxIcon from '@material-ui/icons/Inbox';
+import DraftsIcon from '@material-ui/icons/Drafts';
+
+const styles = (theme: Theme) =>
+  createStyles({
+    root: {
+      width: '100%',
+      maxWidth: 360,
+      backgroundColor: theme.palette.background.paper,
+    },
+  });
+
+function ListItemLink(props: ListItemProps<'a', { button?: true }>) {
+  return <ListItem button component="a" {...props} />;
```

Comment:
@pelotom I guess this is somewhat better than before. It's not truly accurate because I would not be able to use `<ListItemLink component="span" />` but I guess the tradeoff is somewhat ok. I think it's even good to forbid this because it limits the amount of places people would change what DOM node is rendered.

`ListItemProps<'a', { button?: true }>` is however accurate for `<ListItem {...props} button component="a"/>`.

I tried it with union props but gave up at some point. `ListItem` is just a monster when it comes to "what DOM node will be rendered?". I'll look into a simpler component and see if we can simplify those types with union props. I'm suspecting that this will still cause implicit `any` in callbacks because type inference works differently in JSX.